File: October2022/C++/AddOneRowtoTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */


// https://leetcode.com/problems/add-one-row-to-tree/
// TC O(n)
// SC O(height of Tree)

class Solution {
private:
    void insertNewRow(TreeNode*root,int val,int depth,int currDepth){
        
        if(root==NULL){
            return;
        }
        // Adding nodes to current depth
        if(currDepth == (depth-1)){
            TreeNode*currRight = root->right;
            TreeNode*currLeft = root->left;
            root->right = new TreeNode(val);
            root->left = new TreeNode(val);
            root->right->right = currRight;
            root->left->left = currLeft;
            return;
        }
        
        // Searching for required depth
        insertNewRow(root->left,val,depth,currDepth+1);
         insertNewRow(root->right,val,depth,currDepth+1);
    }
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
        
        //  Handling Edge Case
        if(depth==1){
            TreeNode* newRoot = new TreeNode(val);
            newRoot->left = root;
            return newRoot;
        }
        // Function to add Row at given depth
        insertNewRow(root,val,depth,1);
        
        return root;
    }
};

File: October2022/Java/AddOneRowToTree.java
// @saorav21994

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        
        if (depth == 1) {       // make root of tree as left child of new val node (as stated in question)
            TreeNode newNode = new TreeNode(val);
            newNode.left = root;
            return newNode;
        }
        
        // else bfs traversal till (depth-1) level
        
        depth -= 1;
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        
        while (depth-- > 1) {
            
            int size = queue.size();
            while (size-- > 0) {
                
                TreeNode cur = queue.poll();
                if (cur.left != null)
                    queue.offer(cur.left);
                if (cur.right != null)
                    queue.offer(cur.right);
                
            }
            
        }
        
        while (!queue.isEmpty()) 
        {
            
            TreeNode cur = queue.poll();
            
            TreeNode newNodeLeft = new TreeNode(val);
            newNodeLeft.left = cur.left;
            cur.left = newNodeLeft;
            
            TreeNode newNodeRight = new TreeNode(val);
            newNodeRight.right = cur.right;
            cur.right = newNodeRight;
                        
        }
        
        return root;
        
    }
}
